storage/remote/chunked.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package remote
+
+import (
+	"bufio"
+	"encoding/binary"
+	"github.com/gogo/protobuf/proto"
+	"io"
+	"net/http"
+)
+
+// ChunkedWriter is an io.Writer wrapper that allows streaming by adding uvarint delimiter before each write in a form
+// of length of the corresponded byte array.
+type ChunkedWriter struct {
+	writer  io.Writer
+	flusher http.Flusher
+}
+
+// NewChunkedWriter constructs a ChunkedWriter.
+func NewChunkedWriter(w io.Writer, f http.Flusher) *ChunkedWriter {
+	return &ChunkedWriter{writer: w, flusher: f}
+}
+
+// Write writes given bytes to the stream. It adds uvarint delimiter before each message.
```

Comment:
Should we add a checksum while we're at it?

Reply:
I don't have strong opinions here; it adds complexity, OTOH TCP checksums are pretty useless.  I'm inclined to say not right now, mainly as we have the ability to introduce a new framing protocol with checksums in the future.

Reply:
isn't that done on [TCP level](https://locklessinc.com/articles/tcp_checksum/)?

Reply:
As Tom says, TCP checksums are pretty useless.